Refactor test case structure and improve error messages

ryan-gang opened the pull request:
This PR refactors how we handle pattern matching in test cases to provide better error messages and cleaner code organization.

Key changes:

- Split SingleLineOutputTestCase into three specialized test case types:

- SingleLineExactMatchTestCase - For exact string matches with colored diff output

- SingleLinePatternMatchTestCase - For regex pattern matching with simple diff output

- Base singleLineOutputTestCase - Internal implementation shared by both

- Improved error message formatting:

- Added color coding (green for expected, red for actual output)

- Better visual alignment and readability

- More consistent error message structure

- Code cleanup:

- Removed duplicate regexp imports across multiple files

- Simplified pattern matching logic

- Made the base test case type private since it's an implementation detail

- Dependencies:

- Promoted github.com/fatih/color from indirect to direct dependency for colored error messages

These changes make test failures more readable and maintainable while reducing code duplication. The colored output will help developers quickly identify differences between expected and actual output.

No functional changes to the testing behavior - this is purely a refactoring for better developer experience.

rohitpaulk commented:
<img width="506" alt="Screenshot 2024-11-22 at 16 26 41" src="https://github.com/user-attachments/assets/9606c9d5-68e8-4c8d-bcda-5df908bf7e3c">

rohitpaulk commented on internal/test_cases/single_line_exact_match_output_test_case.go:
```diff
@@ -0,0 +1,64 @@
+
+}
+
+func BuildExactMatchValidator(pattern string, simplifiedPatternExplanation string, logger *logger.Logger) func([]byte) error {
```
@ryan-gang this can be private now, it isn't part of the module's interface no?

rohitpaulk commented on internal/test_cases/single_line_exact_match_output_test_case.go:
```diff
@@ -0,0 +1,64 @@
+	Command string
+
+	// ExpectedPattern is the regex that is evaluated against the command's output.
+	ExpectedPattern string
```
@ryan-gang for an exact match, why do we need a pattern? Can't we just pass a string?

rohitpaulk commented on internal/test_cases/single_line_pattern_match_output_test_case.go:
```diff
@@ -0,0 +1,46 @@
+
+}
+
+func BuildPatternMatchValidator(pattern string, simplifiedPatternExplanation string) func([]byte) error {
```
Don't think this needs to be exported

rohitpaulk reviewed (approved):
Not super clear to me how we're handling exact matches - looks like we're still using patterns. Will clarify on today's call! Good to merge for now if this is blocking something else

ryan-gang commented:
Yep, we had a couple of different testcases for exact regex matches `(^foo$)` and exact string matches. 
Refactored them to be a single implementation with the option to pass either one. 
Implemented this here: https://github.com/codecrafters-io/shell-tester/commit/7cc16028d22438a284c041151a6df779d3786c64
